File: SOC/src/cc/game/soc/core/PlayerRandom.java
package cc.game.soc.core;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import cc.lib.game.Utils;

/**
 * An implementation of SOCPlayer that returns random results.
 * 
 * @author ccaron
 *
 */
public class PlayerRandom extends Player {

    public PlayerRandom() {
    }

    private <T> T pickRandom(Collection<T> options) {
        Iterator<T> it = options.iterator();
        int num = Utils.rand() % options.size();
        for (int i=0; i<num-1; i++)
            it.next();
        return it.next();
    }
    
	@Override
	public MoveType chooseMove(SOC soc, Collection<MoveType> moves) {
		return pickRandom(moves);
	}

	@Override
	public Integer chooseVertex(SOC soc, Collection<Integer> vertexIndices, VertexChoice mode, Integer knightToMove) {
		return pickRandom(vertexIndices);
	}

	@Override
	public Integer chooseRoute(SOC soc, Collection<Integer> routeIndices, RouteChoice mode, Integer shipToMove) {
		return pickRandom(routeIndices);
	}

	@Override
	public RouteChoiceType chooseRouteType(SOC soc) {
		return RouteChoiceType.values()[Utils.rand() % RouteChoiceType.values().length];
	}

	@Override
	public Integer chooseTile(SOC soc, Collection<Integer> tileIndices, TileChoice mode) {
		return pickRandom(tileIndices);
	}

	@Override
	public Trade chooseTradeOption(SOC soc, Collection<Trade> trades) {
		return pickRandom(trades);
	}

	@Override
	public Integer choosePlayer(SOC soc, Collection<Integer> playerOptions, PlayerChoice mode) {
		return pickRandom(playerOptions);
	}

	@Override
	public Card chooseCard(SOC soc, Collection<Card> cards, CardChoice mode) {
		return pickRandom(cards);
	}

	@Override
	public <T extends Enum<T>> T chooseEnum(SOC soc, EnumChoice mode, T [] values) {
		return values[Utils.rand() % values.length];
	}

	@Override
	public boolean setDice(SOC soc, List<Dice> die, int num) {
		for (int i=0; i<num; i++) {
			die.get(i).roll();
		}
		return true;
	}
    
    
}

File: SOC/src/cc/game/soc/core/PlayerRandom.kt
package cc.game.soc.core

import cc.lib.game.Utils

/**
 * An implementation of SOCPlayer that returns random results.
 *
 * @author ccaron
 */
class PlayerRandom : Player() {
	private fun <T> pickRandom(options: Collection<T>): T {
		val it = options.iterator()
		val num = Utils.rand() % options.size
		for (i in 0 until num - 1) it.next()
		return it.next()
	}

	override fun chooseMove(soc: SOC, moves: Collection<MoveType>): MoveType? {
		return pickRandom(moves)
	}

	override fun chooseVertex(soc: SOC, vertexIndices: Collection<Int>, mode: VertexChoice, knightToMove: Int?): Int? {
		return pickRandom(vertexIndices)
	}

	override fun chooseRoute(soc: SOC, routeIndices: Collection<Int>, mode: RouteChoice, shipToMove: Int?): Int? {
		return pickRandom(routeIndices)
	}

	override fun chooseRouteType(soc: SOC): RouteChoiceType? {
		return RouteChoiceType.values()[Utils.rand() % RouteChoiceType.values().size]
	}

	override fun chooseTile(soc: SOC, tileIndices: Collection<Int>, mode: TileChoice): Int? {
		return pickRandom(tileIndices)
	}

	override fun chooseTradeOption(soc: SOC, trades: Collection<Trade>): Trade? {
		return pickRandom(trades)
	}

	override fun choosePlayer(soc: SOC, playerOptions: Collection<Int>, mode: PlayerChoice): Int? {
		return pickRandom(playerOptions)
	}

	override fun chooseCard(soc: SOC, cards: Collection<Card>, mode: CardChoice): Card? {
		return pickRandom(cards)
	}

	override fun <T : Enum<T>> chooseEnum(soc: SOC, mode: EnumChoice, values: Array<T>): T? {
		return values[Utils.rand() % values.size]
	}

	override fun setDice(soc: SOC, die: List<Dice>, num: Int): Boolean {
		for (i in 0 until num) {
			die[i].roll()
		}
		return true
	}
}